### PR TITLE
Feature/cube corners

### LINKED ISSUE #541: Getting grid-box corners out of HISTORY
For seamless plotting of data on curvlinear grids with functions like matplotlib's [`pcolormesh(X, Y, C)`](https://matplotlib.org/3.3.1/api/_as_gen/matplotlib.pyplot.pcolormesh.html), the grid-box corner coordinates are needed. Would it be possible to get corner coordinates in diagnostics out of HISTORY? Functions like `pcolormesh()` expect `X` and `Y` to be size `N+1` if the shape of `C` is `(N, N)`, so corners with duplicates (between faces) would be preferred.

Following up on geoschem/GCHPctm#38 and this comment by @tclune 
> I have discussed with @atrayanov and @ben-nasa and they think this will be straightforward. Please open a ticket in MAPL and assign to @ben-nasa.
> 
> Basically we already have a per-grid factory with a method that specifies metadata. That routine would need a few more lines to add metadata for corners. And then the generic layer that sends data to the I/O server needs a few lines to extract the actual values of the coordinates from the ESMF Grid object. The biggest concern is the handling of the duplicate corners, and the MAPL ticket you would open is the best place to discuss that.

### PR DESCRIPTION
**Pull request by bena-nasa**
Fixes #541 
This adds the corners to the cubed-sphere output. I unfortunately had to modify all the grid factories since I needed to make a new routine to get the correct file bounds in the edge case and this is deferred from the abstract factory.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Trivial change (affects only documentation or cleanup)

## Checklist:

- [X] I have tested this change with a run of GEOSgcm (if non-trivial)
- [X] I have added one of the required labels (0 diff, 0 diff trivial, 0 diff structural, non 0-diff)
- [ ] I have updated the CHANGELOG.md accordingly following the style of [Keep a Changelog](https://keepachangelog.com/en/1.0.0/#how)

### CODE REVIEW

File: base/MAPL_AbstractGridFactory.F90 [hunk 1]
```diff
@@ -183,6 +184,18 @@ subroutine generate_file_bounds(this,grid,local_start,global_start,global_count,
          
       end subroutine generate_file_bounds
 
+      subroutine generate_file_edge_bounds(this,grid,local_start,global_start,global_count,rc)
```
Comment (tclune) [inline base/MAPL_AbstractGridFactory.F90:187]: Do you mean edge here?  Or corner?

File: base/MAPL_AbstractGridFactory.F90 [hunk 2]
```diff
@@ -183,6 +184,18 @@ subroutine generate_file_bounds(this,grid,local_start,global_start,global_count,
...
+         use esmf
+         import AbstractGridFactory
+         class (AbstractGridFactory), intent(inout) :: this
+         type(ESMF_Grid), intent(inout)      :: grid
+         integer, allocatable, intent(inout) :: local_start(:)
+         integer, allocatable, intent(inout) :: global_start(:)
+         integer, allocatable, intent(inout) :: global_count(:)
```
Comment (tclune) [inline base/MAPL_AbstractGridFactory.F90:194]: I believe these should be `intent(out)`, not `inout`.    With allocatable, the usual concerns that often force us to conservatively use the broader INTENT are not a concern.

File: base/MAPL_newCFIO.F90 [hunk 3]
```diff
@@ -645,6 +646,37 @@ subroutine stage2DLatLon(this, fileName, oClients, rc)
...
+        _VERIFY(status)
+
+        call factory%generate_file_edge_bounds(this%output_grid,LocalStart,GlobalStart,GlobalCount,rc=status)
+        _VERIFY(status)
+        call ESMF_GridGetCoord(this%output_grid, localDE=0, coordDim=1, &
+        staggerloc=ESMF_STAGGERLOC_CORNER, &
+        farrayPtr=ptr2d, rc=status)
+        _VERIFY(STATUS)
+        if (.not.allocated(this%corner_lons)) allocate(this%corner_lons(size(ptr2d,1),size(ptr2d,2)))
+        this%corner_lons=ptr2d*MAPL_RADIANS_TO_DEGREES
```
Comment (tclune) [inline base/MAPL_newCFIO.F90:667]: I think you can do this more elegantly.  No need for the explicit allocate.   Just
```
this%corner_lons = ptr2d * MAPL_RADIANS_TO_DEGREES
```

File: base/MAPL_newCFIO.F90 [hunk 4]
```diff
@@ -645,6 +646,37 @@ subroutine stage2DLatLon(this, fileName, oClients, rc)
...
+        ref = ArrayReference(this%corner_lons)
+         call oClients%collective_stage_data(this%write_collection_id,trim(filename),'corner_lons', &
+              ref,start=localStart, global_start=GlobalStart, global_count=GlobalCount)
+        call ESMF_GridGetCoord(this%output_grid, localDE=0, coordDim=2, &
+        staggerloc=ESMF_STAGGERLOC_CORNER, &
+        farrayPtr=ptr2d, rc=status)
+        _VERIFY(STATUS)
+        if (.not.allocated(this%corner_lats)) allocate(this%corner_lats(size(ptr2d,1),size(ptr2d,2)))
+        this%corner_lats=ptr2d*MAPL_RADIANS_TO_DEGREES
```
Comment (tclune) [inline base/MAPL_newCFIO.F90:676]: As above - no need for the explicit allocate statement.

**Review by tclune — CHANGES_REQUESTED** [review]
Mostly minor changes requested.  See the inline comments.   But edge vs corner is important to get right.

File: base/MAPL_newCFIO.F90 (hunk 3 above)
Comment (bena-nasa) [reply base/MAPL_newCFIO.F90:667]: Is '=' overloaded to do the allocation and match the shape/bounds of the right hand side if the left hand side is not allocated?

Comment (tclune) [reply base/MAPL_newCFIO.F90:667]: "Overloaded" is the wrong term here.  In F2003, _intrinsic_ assignment with allocatable LHS is required to allocate (or reallocated) the LHS unless it already matches the shape of the RHS.   Exactly suited to this case.

F2008 went a step further and included reallocating if the LHS is polymorphic and not of the same type as the rhs.

Comment (bena-nasa) [reply base/MAPL_newCFIO.F90:665]: Make the requested changes including an updating the mockgridfactory so hopefully the MAPL unit tests pass this time.

**Comment by bena-nasa** [comment]
Made requested changes including update to mockgridfactory so hopefully units tests pass this time.

**Review by tclune — APPROVED** [review]

**Pull request merged**